### ISSUE TITLE
Rework string literal handling for safety

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#639]
Solidity states that all strings are UTF8. Even though it doesn't offer internal methods for strings, all the libraries written for it are based on this assumption.

Contrary to this, the hex escape works as in C: `\x80` means the byte `0x80`. In Javascript (and others with strict UTF8 compliance) it means the UTF8 encoded Unicode code point of `0080`: `0xc2 0x80`.

The main problem with the current behaviour is that it creates invalid UTF8 strings. This is the real reason why #603 is an issue.

I think it should behave like Javascript and include better support for assigning binary data to the `bytes` (and `bytesNN`) type, such as the one proposed in #479.

**Comment by chriseth** [comment]
The problem with string literals is that they can either be `string`s, `bytes`s or `bytes32` (or even shorter). What about parsing `\x80` as the byte `0x80` and `\u0080` as the unicode code point `0x0080` encoded in utf-8?

**Comment by axic** [comment]
Is there a way to have 2 sub types of `StringLiteral`, one used in `string`s, while the other used in `bytes*`?

That should make it work cleanly.

**Comment by chriseth** [comment]
Yes, but that would require the user to type them differently, e.g. something like `s"abc"` or `b"abc"`. I would say that using `\x` and `\u` to denote different things is also clean.

**Comment by axic** [comment]
Probably it wouldn't hurt leaving it `"abc"` for `string` and introducing something else for `bytes`, since then it could also enjoy #479 implemented a nicer way.

Additionally to solve the above issue I think we could do the following too:
a) Introduce a warning when a non-UTF8 string literal is encoded. (As `string` is defined as strictly UTF8.)
b) Change `StringLiteral::toString()` so that it outputs `"non-UTF8 literal string"` or `"invalid literal string"` - this would fix #603

**Comment by chriseth** [comment]
How would you propose to signal a bytes literal (or other way to input raw bytes)?

**Comment by axic** [comment]
If the tokenizer cannot use the state (i.e. we are in a bytes assignment sequence) and the solution proposed in #603 is not something feasible, then perhaps using the backtick (or another similar character) might be a good trade off.

```
bytes code = `606060405260188060106000396000f360606040523615600d57600d565b60165b6002565b565b00`;
```

However, reading the previous comment again:

> introducing something else for bytes

Here I've meant that if possible, bytes could have the `b` prefix:

```
bytes code = b"Hello World \xFF";
```

**Comment by chriseth** [comment]
@axic I don't find a proposed solution in #603. Yes, I was talking about something like `b"..."` to distinguish between string literals and byte array literals. If possible, we should also allow things like `{0x12, 0x34, 0x11, ...}` - but that is something else.

So basically as a summary: I proposed to parse `\x12` as raw byte and `\u0012` as unicode (then encoded as utf-8). You wanted to signal a warning if the parsed raw bytes do not represent valid utf-8. Note that this also means that we force the source code encoding to utf-8 and parse actual raw utf-8 bytes as that.

If we signal a warning, we have to provide a way to "silence" it and tell the compiler "Yes, I really want a raw byte array, even if it is not utf-8". This silencing mechanism could be to prefix the literal with a `b`. Are there other solutions? Perhaps `r"..."`?

As backticks have a special meaning in ecmascript, I would not use them.

Oh and I think a context-aware literal parser/scanner is no the best idea :-)

**Comment by axic** [comment]
Sorry I've meant #479 and not #603 - but let's just put these aside for a moment.

Solidity (and the ABI) documentation states that strings **must be** UTF8. It would be nice to take that seriously to avoid any possible issues down the line.

If I understand it correctly, the tokenizer needs uniquely identifiable tokens to do the parsing/validation - hence we cannot postpone this validation to when the parsed data is assigned  to the appropriate type.

My motivation regarding escapes/strings/bytes:
- Fix #603 
- Support #479 in some way (and perhaps support other ways you've proposed, such as `{0x12, 0x34, 0x11, ...}`)
- Ensure that strings are UTF8

## 

Summarising this issue:
- `""` should continue to work for both `string` and `bytes`
- `\xXX` should work for escaping any hex value
- `\uXXXX` should take a valid Unicode codepoint and encode an UTF8 sequence
- `StringLiteral:toString()` should not return an invalid UTF8 sequence as that produces an invalid JSON compiler output. It should instead return `non-UTF8 string literal` or `invalid string literal`
- When an invalid UTF8 sequence is detected in a `string` assignment, the compiler should throw a warning, unless the literal was enclosed in`b""`

Does the above makes sense?

The last point is the only one I'm not sure how to actually implement.

**Comment by chriseth** [comment]
Makes sense! As the ast output also has the source reference, it should be easy to retrieve the actual literal, so the `StringLiteralType::toString()` function could return `string_literal (non-utf8) "abc??def"` for invalid utf-8 data.

Concerning the warning: If you say we only warn at the point of the assignment (is that possible?) - the warning could also be silenced by using `x = bytes("abc\xff")`.

**Comment by axic** [comment]
Python uses `b""` for _byte literals_ so that's perhaps :+1: for it.

@chriseth and regarding #479 it is probably better to keep it discussed over there, but the `r` prefix could be a good choice instead of backticks:

```
bytes code = r"60606040";
```

This should be possible with the tokenizer, right?

**Comment by chriseth** [comment]
As we have `hex"..."`, we could also use `raw"..."` ('r' is a bit too short, I would say), although if possible, I would prefer `bytes("abc")`.

**Comment by chriseth** [comment]
What is left to do here, @axic?

**Comment by axic** [comment]
Merging #832.

And:

> When an invalid UTF8 sequence is detected in a string assignment, the compiler should throw a warning, unless the literal was enclosed in `b""`

I would reword it though: _When an invalid UTF8 sequence is detected in a string assignment, the compiler should throw a warning._

The helper is available for that (validate utf8).

**Comment by chriseth** [comment]
Closing in favour of https://github.com/ethereum/solidity/issues/883
